[ENH] New estimator - RegularizedVAR

meraldoantonio opened the pull request:
#### Reference Issues/PRs
N/A

#### What does this implement/fix? Explain your changes.
This PR introduces a new `VARReduce` class, which implements a VAR-like forecaster that educes multivariate time series to tabular regression and trains a regressor.
 Any scikit-learn compatible regressor can be used; by default, `LinearRegression`  is used, making it behave like a traditional VAR. 
Users can specify other regressors such as `ElasticNet` to introduce regularization and potentially enhance performance with large datasets or in the presence of multicollinearity.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
Correctness of implementation

#### Did you add any tests for the change?
Yes, I've tested the fitted parameters and forecasting results of this new class versus a 'normal' VAR from `statsmodels` on a synthetic multivariate data. The results, which are shown in an example notebook, are almost identical. 

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

fkiraly commented on sktime/forecasting/regularized_var.py:
```diff
@@ -0,0 +1,224 @@
+    _tags = {
+        "scitype:y": "both",
+        "authors": ["meraldoantonio"],
+        "python_dependencies": "cvxpy",
```
the model will not be tested in CI if `cvxpy` is not present in the env - so you need to add it at least in the `all_extras` dep set in `pyproject,toml` to ensure test coverage

meraldoantonio replied:
Noted! I've changed the logic of the forecaster to simply use a user-defined regressor for fitting, so I removed all cvxpy references

fkiraly commented on sktime/forecasting/regularized_var.py:
```diff
@@ -0,0 +1,224 @@
+
+        Returns
+        -------
+        X : np.ndarray
```
I think exact sizes and a description of what this does would be very helpful for a reviewer

meraldoantonio replied:
Noted, added shapes of the returned array and description of the function

fkiraly commented on sktime/forecasting/regularized_var.py:
```diff
@@ -0,0 +1,224 @@
+            + self.L2_penalty * cp.norm(coefficients, "fro")
+        )
+        problem = cp.Problem(objective)
+        problem.solve()
```
this is the same as `sklearn` `ElasticNet`, from a math perspective, so questions (I have no clear opinion yet):

* use `sklearn` `ElasticNet` instead?
* should there be a "regressor" instead in the parameter, defaulting to `ElasticNet`, and the user can choose anything else `sklearn` compatible? This could be VARreduce or similar

meraldoantonio replied:
That makes a lot of sense! I've reworked the module; now instead of using `cvxpy`, users can specify a regressor instead. 
I chose `LinearRegression` as the default as this will make `VARReduce` identical to `VAR`, which might avoid confusion.

fkiraly replied:
yes, good choice of default, better than my suggestion!
And users can use `ElasticNet` if they also want L1 penalty.

fkiraly commented on sktime/forecasting/regularized_var.py:
```diff
@@ -0,0 +1,224 @@
+        The L1 regularization penalty.
+    L2_penalty : float, optional (default=0.0)
+        The L2 regularization penalty.
+    """
```
do you have a sci reference that recommends fitting VAR like this?

meraldoantonio replied:
I was looking at the following papers that regularize VAR using L1/L2:
https://arxiv.org/abs/2105.00860
https://strathprints.strath.ac.uk/57829/1/Cavalcante_etal_WE2016_LASSO_vector_autoregression_structures_for_very_short_term_wind_power_forecasting.pdf

fkiraly replied:
thanks! Just to avoid me sifting through pages and pages of dense math if you have already done that, can you point to the place where this specific algorithm (tabularization and then minimization) is suggested?

meraldoantonio replied:
Actually, the papers I found didn't explicitly suggest the specific algorithm of tabularization followed by minimization. I couldn't find any papers that did. The two papers I reviewed simply described the use of L1/L2 regularization in LASSO and their advantages without going through tabularization. However, the algorithm we used is mathematically equivalent to a "direct optimization, such as the previous version I implemented using CVXPY.

For instance, Section 2 of [this paper](https://strathprints.strath.ac.uk/57829/1/Cavalcante_etal_WE2016_LASSO_vector_autoregression_structures_for_very_short_term_wind_power_forecasting.pdf) describes the motivation for using LASSO VAR:

---
The VAR model allows a simultaneous forecast of the wind generation in several neighboring sites combining time
series information. However, forecasting with VAR models may be intractable for high dimensional data since the nonsparse coefficients matrix grows quadratically with the number of series included in the model. In order to overcome this limitation, in [28] it is proposed the combination of LASSO and VAR frameworks, which is further explored in this paper for very short-term forecasting of wind power.
---

Similarly, Section 2 of [this paper](https://arxiv.org/pdf/2105.00860) provides a discussion of the ridge penalty and the ridge VAR estimator.

meraldoantonio commented:
## Summary of Changes

1. **Changed the Name of the Module and Class**
2. **Reworked the Fitting Logic**
   - The fitting process no longer uses `cvxpy`.
   - The new fitting process includes:
     - tabularizing fitted data
     - fitting the data to a user-defined `scikit-learn` compatible regressor
     - using `LinearRegression` as the default regressor, ensuring that `VARReduce` behaves identically to `VAR` under default conditions.

3. **Created an Example Notebook**
   - Developed a notebook to compare `VARReduce` with `VAR`.
   - Verified that both models behave the same under default conditions.
   - Demonstrated how `VARReduce` is beneficial for handling multivariate time series with numerous predictors.

fkiraly reviewed (changes requested):
Almost done, there are some minor non-conformances.

* `get_test_params` should cover the default setting, without passing `regressor`, as a second parameter set
* this would have highlighted that `self.regressor` is mutated. Instead, we need to `clone` the regressor, and use `self.regressor_` as a variable

Regarding the reference, we should either point to a reference that has precisely this (very reduction-like) algorithm, or explain what it does in the docstring.

fkiraly reviewed (changes requested):
I've further reviewed the estimator.

* I've made some API conformance changes, see above. Please review the last commit
* I added the check against `statsmodels` `VAR`
* This is not ready to merge though, since estimators that do not have the `coefficients` attribute will fail to run. This is because `_predict` manually computes the prediction, instead applying `model.predict`. I think we should change the logic to use `regressor_.predict` rather than manually reimplementing predict.

fkiraly commented:
(code formatting issues, missing imports)

meraldoantonio commented:
Changing the `.regressor` to `.regressor_` gave me an error during fitting
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], [line 6](vscode-notebook-cell:?execution_count=4&line=6)
      [4](vscode-notebook-cell:?execution_count=4&line=4) # Instantiate and fit the custom VARReduce model
      [5](vscode-notebook-cell:?execution_count=4&line=5) varreduce_model = VARReduce(lags=LAGS)  # no regressor passed, by default, LinearRegressor() will be used
----> [6](vscode-notebook-cell:?execution_count=4&line=6) varreduce_model.fit(df_train)
      [8](vscode-notebook-cell:?execution_count=4&line=8) # Fit the VAR model comparison
      [9](vscode-notebook-cell:?execution_count=4&line=9) var_model = VAR(endog=df_train)

File ~/Desktop/sktime_dev/sktime/sktime/forecasting/base/_base.py:373, in BaseForecaster.fit(self, y, X, fh)
    [370](https://file+.vscode-resource.vscode-cdn.net/Users/meraldo/Desktop/sktime_dev/sktime/examples/forecasting/~/Desktop/sktime_dev/sktime/sktime/forecasting/base/_base.py:370) assert y is not None, "y cannot be None, but found None"
    [372](https://file+.vscode-resource.vscode-cdn.net/Users/meraldo/Desktop/sktime_dev/sktime/examples/forecasting/~/Desktop/sktime_dev/sktime/sktime/forecasting/base/_base.py:372) # if fit is called, estimator is reset, including fitted state
--> [373](https://file+.vscode-resource.vscode-cdn.net/Users/meraldo/Desktop/sktime_dev/sktime/examples/forecasting/~/Desktop/sktime_dev/sktime/sktime/forecasting/base/_base.py:373) self.reset()
    [375](https://file+.vscode-resource.vscode-cdn.net/Users/meraldo/Desktop/sktime_dev/sktime/examples/forecasting/~/Desktop/sktime_dev/sktime/sktime/forecasting/base/_base.py:375) # check and convert X/y
    [376](https://file+.vscode-resource.vscode-cdn.net/Users/meraldo/Desktop/sktime_dev/sktime/examples/forecasting/~/Desktop/sktime_dev/sktime/sktime/forecasting/base/_base.py:376) X_inner, y_inner = self._check_X_y(X=X, y=y)

File /opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:130, in BaseObject.reset(self)
    [108](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:108) """Reset the object to a clean post-init state.
    [109](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:109) 
    [110](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:110) Using reset, runs __init__ with current values of hyper-parameters
   (...)
    [127](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:127) call, self is equal in value to `type(self)(**self.get_params(deep=False))`
    [128](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:128) """
    [129](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:129) # retrieve parameters to copy them later
--> [130](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:130) params = self.get_params(deep=False)
...
--> [278](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:278)     params = {key: getattr(self, key) for key in self.get_param_names()}
    [280](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:280)     if deep:
    [281](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/site-packages/skbase/base/_base.py:281)         deep_params = {}

AttributeError: 'VARReduce' object has no attribute 'regressor'

fkiraly commented:
You need to set *both* `regressor` and `regressor_`.

The former should always be identical with what was passed to `__init__` as the `regressor` arg, whereas the latter you can use as you want.

fkiraly commented:
code formatting fails - I recommend you set up precommit as described here:
https://www.sktime.net/en/latest/developer_guide/coding_standards.html#setting-up-local-code-quality-checks

fkiraly commented:
success :-)

meraldoantonio commented:
> success :-)

Yes, finally :) Not sure why sometimes ruff passes on my local but fails on Github. 
Anyway, the code is ready for review, thank you!

fkiraly commented on sktime/forecasting/var_reduce.py:
```diff
@@ -0,0 +1,296 @@
+
+        # Concatenate the list of Series into the final DataFrame
+        y_pred = pd.concat(y_pred)
+        y_pred.index = fh.to_absolute_index()
```
either you need to pass the `self.cutoff` as an arg, or use the methog `get_expeced_pred_idx` (which also wants the cutoff)

meraldoantonio replied:
(Addressed and resolved)

fkiraly commented:
looks good!

meraldoantonio commented:
> looks good!

Thanks! I've switched the PR back to 'Ready for Review'

fkiraly commented:
Note to reviewers - failures are unrelated and caused by this: https://github.com/sktime/sktime/issues/6975

fkiraly commented on sktime/forecasting/var_reduce.py:
```diff
@@ -0,0 +1,334 @@
+        self.num_series = None
+        super().__init__()
+
+    def prepare_for_fit(self, data, return_as_ndarray=True):
```
can you kindly make this private? Just add an underscore to the name.
(because we would not expect users to call this)

meraldoantonio replied:
(Resolved)

fkiraly commented on sktime/forecasting/var_reduce.py:
```diff
@@ -0,0 +1,334 @@
+        else:
+            return X, Y
+
+    def prepare_for_predict(self, data, return_as_ndarray=True):
```
same, I think this should be private

meraldoantonio replied:
(Addressed and resolved)

fkiraly reviewed (changes requested):
Great! This is basically done, only smaller things, mostly software structure.

Some blocking change requests:
* I think the methods that we use only internally and do not intend users to call (is this intention correctly understood?) should be made private, i.e., starting with underscore
* similarly, in-sample predictions should be a private attribute
* question: should in-sample predictions not be made *only if they are later requested*? We remember the data as a pointer anyway, in `_y` and `_X`.
* `get_test_params` should test using an estimator which is not the default, and not a linear regressor. How about `RandomForestRegressor`?
* can we separate out the notebook into another PR? I think it is really nice, though it may need some additional iterations. I think it would be best if we split it out. The other PR should depend on the PR that contains the estimator.

A blocking question, but it may not require a change:

* why are we specifying `scitype:y = "both"`, i.e., multivariate capability, if all we do is looping over the output? Should we not, instead, allow use of multivariate sklearn regressors to predict multiple variables at once? See "multioutput regressors" at sklearn.

Not blocking:

* I think the docstring should explain the algorithm better, and give referenes. It might also be important to note the parameter combinations for which the "usual" VAR is recovered, and point out some parameter combinations that specialize to things like L1 regularized VAR or elastic-VAR.

meraldoantonio commented:
Thanks for the check! 
Here are my responses:

> * I think the methods that we use only internally and do not intend users to call (is this intention correctly understood?) should be made private, i.e., starting with underscore
Changed!

> * similarly, in-sample predictions should be a private attribute
Changed!

> * question: should in-sample predictions not be made _only if they are later requested_? We remember the data as a pointer anyway, in `_y` and `_X`, so we could do it in `_predict` and only if we need it.
Changed! Now in-sample predictions are made in predict only if needed.

> * `get_test_params` should test using an estimator which is not the default, and not a linear regressor. How about `RandomForestRegressor`?
Added! I used `RandomForestRegressor` and `SVR`

> * can we separate out the notebook into another PR? I think it is really nice, though it may need some additional iterations. I think it would be best if we split it out. The other PR should depend on the PR that contains the estimator.
Noted - removed the notebook from this PR.

> A blocking question, but it may not require a change:
> 
> * why are we specifying `scitype:y = "both"`, i.e., multivariate capability, if all we do is looping over the output? Should we not, instead, allow use of multivariate sklearn regressors to predict multiple variables at once? See "multioutput regressors" at sklearn.
My understanding is that setting scitype:y = "both" indicates that y can be either univariate or multivariate, which is indeed the case for our model. Please correct me if I’m wrong!

I just learned about multioutput regressors in `scikit-learn`. It seems feasible to support multioutput regressors in our model. To implement this, we would first need to check if the provided regressor supports multioutput. If it does, we could fit the model to all Y's simultaneously. If the regressor does not support multioutput, we would continue with the current approach, fitting each Y individually. Should I implement this?

> 
> Not blocking:
> 
> * I think the docstring should explain the algorithm better, and give referenes. It might also be important to note the parameter combinations for which the "usual" VAR is recovered, and point out some parameter combinations that specialize to things like L1 regularized VAR or elastic-VAR.
I modified the docstring, hopefully it looks better now. 
Unfortunately, I can't find any reference that explicitly describes our model, i.e. tabularization + regularization.

fkiraly commented:
> Noted - removed the notebook from this PR.

But you should still add it to another PR, it was a nice notebook!

> If it does, we could fit the model to all Y's simultaneously. If the regressor does not support multioutput, we would continue with the current approach, fitting each Y individually. Should I implement this?

That would work - it is one of two solutions to the current problem, the other one would be just turning the tag to `scitype:y = "univariate"`, but I think this one is nicer, although a bit more work.

> I modified the docstring, hopefully it looks better now.

yes, great!

> Unfortunately, I can't find any reference that explicitly describes our model, i.e. tabularization + regularization.

Then we may want to write one :-)

All other points addressed imo.

fkiraly commented on sktime/forecasting/var_reduce.py:
```diff
@@ -0,0 +1,369 @@
+    be used to introduce regularization and/or non-linearity.
+
+    For example:
+    - VARReduce(regressor = Ridge()) is equivalent to VAR with L2 regularization;
```
minor remark: does this display properly, or do we need to add spaces around the bullet points?

meraldoantonio replied:
Yes I added a tab before the dash

meraldoantonio replied:
(Addressed and resolved)

fkiraly commented on sktime/forecasting/var_reduce.py:
```diff
@@ -0,0 +1,369 @@
+
+    Fitting proceeds in two steps:
+    1. Tabularization: For each time step and each time series within the
+       multivariate input data, lagged values `X` are generated.
```
minor remark: a short mathematical or pictorial description would be useful to the reader

meraldoantonio replied:
Agreed; I added pictorial description in the docstring

meraldoantonio commented:
> > Noted - removed the notebook from this PR.
> 
> But you should still add it to another PR, it was a nice notebook!
> 

Sure! I've opened a child PR for the notebook.


> > If it does, we could fit the model to all Y's simultaneously. If the regressor does not support multioutput, we would continue with the current approach, fitting each Y individually. Should I implement this?
> 
> That would work - it is one of two solutions to the current problem, the other one would be just turning the tag to `scitype:y = "univariate"`, but I think this one is nicer, although a bit more work.
> 

Sure; I added the support for multioutput.


> > I modified the docstring, hopefully it looks better now.
> 
> yes, great!
> 
> > Unfortunately, I can't find any reference that explicitly describes our model, i.e. tabularization + regularization.
> 
> Then we may want to write one :-)
> 
>

Yes that sounds like a plan! Let's discuss about it during our 1-on-1!

fkiraly reviewed (changes requested):
All above addressed.

There are new failures related to `reshape` which look genuine, can you kindly debug?

fkiraly commented on sktime/forecasting/var_reduce.py:
```diff
@@ -0,0 +1,445 @@
+        intercepts = np.zeros(num_series)
+
+        # Try multioutput fitting
+        try:
```
try-except to detect a condition is a "bad pattern" - instead, one should try to determine the condition explicitly.

The reason is that try/except is often more general than the condition one actually wants to capture, and genuine exceptions may get caught up without bubbling to the surface.

meraldoantonio replied:
(Addressed and resolved)

fkiraly commented on sktime/forecasting/var_reduce.py:
```diff
@@ -0,0 +1,445 @@
+            # for each target series `y` in `Y`, a separate regressor is
+            # trained using `y` and the common predictors `X`.
+            # TODO: use sklearn's MultiOutputRegressor to unify API
+            for i, var_name in enumerate(var_names):
```
what I would do instead, if `y` is multivariate, is to wrap the regressor in `MultiOutputRegressor`, and write it to `regressor_`.

That way, `regressor_` has no case-by-case differring type, and is always the fitted regressor.

It should also solve the remaining failures related to `get_fitted_params`, which encounters an unfitted `regressor_`.

meraldoantonio replied:
(Addressed and resolved)

fkiraly commented:
looks like the interface uniformization also fixed the `get_fitted_params` issue

fkiraly reviewed (approved):
All addressed, nice estimator!

meraldoantonio commented:
> looks like the interface uniformization also fixed the `get_fitted_params` issue

Yep :-)
I've switched the PR back to 'Ready for Review'

meraldoantonio commented:
> All addressed, nice estimator!

Thanks! I've resolved all conversations... is there anything else I should do before merging?

fkiraly commented:
no, thx
